### DISCUSSION TITLE
Carrier Loss

### DISCUSSION THREAD

**Discussion by matwell-dev** [rsmarples/dhcpcd#65] [Q&A] [answered]
Hi!

I recently had a one-off(?) and probably short-lived (I wasn't in the building at the time, but the hub and Pi are in the same room!) - but apparently 'fatal' - carrier loss on one of my Raspberry Pi's to my wifi. The other one had one at a similar time and seems to have successfully reconnected. Restarting the dhcpcd seems to have reconnected and now all is fine. I am keen to understand what detects carrier-loss - is it DHCPCD itself? It certainly reported it? And also, whether there's a configuration option to specify retries - number, period etc. Again, is this part of the daemon - I can't find anything about it, or should I be looking somewhere else?

**Versions:**
dhcpcd 6.11.5
Copyright (c) 2006-2016 Roy Marples
Compiled in features: INET IPv4LL INET6 DHCPv6 AUTH

Linux rasp-pi2 4.19.66-v7+ #1253 SMP Thu Aug 15 11:49:46 BST 2019 armv7l GNU/Linux

I took a look at the latest source-code, which appears to be several generations on and features a 'roaming' facility which looks like it is more wifi and its more ethereal nature friendly?

Thoughts and comments welcome!

Cheers

Mark

**Comment by rsmarples** [comment, marked as answer]
Hi!

carrier loss / gain is detected by the kernel driver for the interface and then advertised via netlink(7) RTM_NEWLINK message.
dhcpcd has changed over the years on what it reads from this to work out carrier loss / gain and the kernel has gained features in this area as well, such as partial carrier such as wireless roaming.

dhcpcd reports these events and certainly 6.11.5 is an old version where the handling throughout has seen many improvements.
Certainly that version is so old it might interpret the roaming features now reported by the kernel as incorrect and not recover from it.
I would upgrade for sure.

**Comment by rat-dev** [comment]
Hi there!

I recently started to use dhpcd and was wondering about when hooks are actually called, if it comes to CARRIER/NOCARRIER state changes:
What I observe is on dhcpcd startup that hooks are indeed called after PREINIT for CARRIER if cable connected / NOCARRIER if cable not connected and subsequently CARRIER when the cable is plugged in. Then dhcpcd proceeds, in my case to STATIC since I have  a static config, so far so good.

However subsequent cable pulling/re-plugging is detected by dhcpcd and route setting/deliting is done correctly, however, the hooks arent called for those carrier events. I was wondering if this is the expected behaviour? 
dhcpcd version is 9.4.1

Thx and regards,

Harry

**Reply by rsmarples** [reply]
If dhcpcd is detecting cable pulling/re-plugging then it is also ruunning the hooks as it's literally the next step after logging the detection.
https://github.com/NetworkConfiguration/dhcpcd/blob/master/src/dhcpcd.c#L742

Can you post a dhcpcd debug log for this please?

**Reply by rat-dev** [reply]
I just looked into the hook file I wrote...looks like my logging was only working partially ): so basically in deed a carrier detection triggers the hooks being called. Sorry for this false positive, my bad!

Greez,

Harry
